client/src/erlang/receiver.zig

```diff
@@ -0,0 +1,311 @@
+    if (item.is_tuple) {
+        try erl.validate(
+            error.decoding_tuple,
+            ei.ei_decode_tuple_header(deserializer.buf.buff, deserializer.index, &size),
+        );
+        if (item.fields.len != size) return error.wrong_tuple_size;
+        inline for (&value) |*elem| {
+            elem.* = try internal_receive_message(@TypeOf(elem.*), allocator, deserializer);
+        }
+    } else {
+        try erl.validate(
+            error.decoding_map,
+            ei.ei_decode_map_header(deserializer.buf.buff, deserializer.index, &size),
+        );
+        const fields = std.meta.fields(T);
+        var present_fields: [fields.len]bool = .{false} ** fields.len;
+        var counter: u32 = 0;
+        if (size > fields.len) return error.too_many_map_entries;
+        for (0..@as(u32, @bitCast(size))) |_| {
+            const key = try receive_atom(deserializer, allocator);
+            inline for (0.., fields) |idx, field| {
+                if (std.mem.eql(u8, field.name, key)) {
+                    const current_field = &@field(value, field.name);
+                    const field_type = @typeInfo(field.type);
+                    if (field_type == .Optional) {
+                        current_field.* = try internal_receive_message(field_type.Optional.child, allocator, deserializer);
+                    } else {
+                        current_field.* = try internal_receive_message(field.type, allocator, deserializer);
+                    }
+                    present_fields[idx] = true;
+                    counter += 1;
+                }
+            }
+        }
+        if (size < counter) return error.too_few_map_entries;
+        var should_error = false;
+        inline for (present_fields, fields) |presence, field| {
+            if (!presence) {
+                if (@typeInfo(field.type) == .Optional) {
+                    const current_field = &@field(value, field.name);
+                    current_field.* = null;
+                } else {
+                    std.debug.print("Missing Field in Struct {s}: {s}\n", .{ @typeName(T), field.name });
+                    should_error = true;
+                }
+            }
+        }
+        if (should_error) return error.missing_field_in_struct;
+    }
+    return value;
+}
+
+inline fn receive_int(comptime T: type, comptime item: std.builtin.Type.Int, deserializer: Deserializer) !T {
+    // TODO: eventually arbitrarily sized integers.
+    var value: T = undefined;
+    if (item.signedness == .signed) {
+        try erl.validate(error.decoding_signed_integer, ei.ei_decode_long(deserializer.buf.buff, deserializer.index, &value));
+    } else {
+        try erl.validate(error.decoding_unsigned_integer, ei.ei_decode_ulong(deserializer.buf.buff, deserializer.index, &value));
+    }
+    return value;
+}
+
+inline fn receive_enum(comptime T: type, comptime item: std.builtin.Type.Enum, allocator: std.mem.Allocator, deserializer: Deserializer) !T {
```

Comment:
I think we actually shouldn't receive an allocator for this one. Enums by definition cannot have nested members, and their size is known at compile time. Furthermore, we also know at compile time the maximum length of all the names of each tag in the enum. We should be able to use a `FixedBufferAllocator`.

Reply:
We are not receiving the allocator anymore, per say. We receive `@This()`.